src/testing/util.ts

```diff
@@ -0,0 +1,81 @@
+interface Waiter {
+  done: boolean;
+  wakeUp: () => void;
+}
+
+/**
+ * Implementation of a Semaphore class to control the number of
+ * concurrent operations for a given task.
+ *
+ * https://github.com/python/cpython/blob/5dc8c84d397110f9edfa56793ad8887b1f176d79/Lib/asyncio/locks.py#L352
+ */
+export class Semaphore {
+  private available: number;
+  private waiters: Set<Waiter>;
+
+  constructor(max: number) {
+    this.available = max;
```

Comment:
What do you think about computing available based on max and # of waiters? That way we don't have to separately track it.

Reply:
we don't actually start adding waiters until available is 0, so i think it needs to be tracked separately: 

```ts
  private async acquire(): Promise<void> {
    if (this.available > 0) {
      this.available--;
      return;
    }
```

Reply:
btw was following this if it's helpful: https://github.com/python/cpython/blob/5dc8c84d397110f9edfa56793ad8887b1f176d79/Lib/asyncio/locks.py#L352